### PR TITLE
Multiple entity support for ExecuteUpdate

### LINKED ISSUE #31349: EntityFrameworkCore 7 SetProperty(avar,avalue) Unable to cast object of type 'System.Linq.Expressions.TypedParameterExpression' to type 'System.Linq.Expressions.LambdaExpression'.
1. Visual Studio 2022
2. .NET 7.0
3. ASP.NET Core Web API Project
4. OS Windows10
5. Microsoft.EntityFrameworkCore 7.0.9

**My code:
1.Error**
```
ADal aDal = new ADal();

Func<AModel, float> p = property => property.Balance;

aDal._databaseContext
.ModelDbSet
.Where(w => w.ID == 5)
.Take(1)
.ExecuteUpdate(e => e.SetProperty(p, 636));
```

Out error (Almost no answer found online):
引发的异常:“System.InvalidCastException”(位于 System.Private.CoreLib.dll 中)

“System.InvalidCastException”类型的未经处理的异常在 System.Private.CoreLib.dll 中发生

Unable to cast object of type 'System.Linq.Expressions.TypedParameterExpression' to type 'System.Linq.Expressions.LambdaExpression'.

**2.Normal (Do not use : Func<AModel, float> p = property => property.Balance;)**
```
ADal aDal = new ADal();

aDal._databaseContext
.ModelDbSet
.Where(w => w.ID == 5)
.Take(1)
.ExecuteUpdate(e => e.SetProperty(p => p.Balance, 636));
```

### LINKED ISSUE #31406: Allow ExecuteUpdate to update properties of multiple queries as long as the map to a single table
Our current ExecuteUpdate logic checks that only one entity type is updated by all the setters; this blocks scenarios such as the following:

### Updating columns mapped to multiple owned entities (and the owner):

```c#
_ = await ctx.Blogs.ExecuteUpdateAsync(s => s
    .SetProperty(b => b.Name, "foo")
    .SetProperty(b => b.BlogDetails.Details, "bar"));
```

(note: this will also apply to complex types)

### Updating columns mapped to a single table within a TPT hierarchy:

```c#
_ = await ctx.SpecialBlogs.ExecuteUpdateAsync(
    s => s.SetProperty(b => b.SpecialProperty, "foo"));
```

### Updating the **main** table of an entity with entity splitting

As long as updates only affect columns on the main table, this now works.

Unfortunately, updates to any of the split tables fail on SQL Server for other, architectural reasons, see #31407. On Sqlite they pass, because a different query shape is generated.

### PR DESCRIPTION
**Pull request by roji**
Closes #31406
Fixes #31349

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesSqlServerTest.cs [hunk 1]
```diff
@@ -971,7 +969,11 @@ public override async Task Update_Where_Distinct_set_constant(bool async)
 UPDATE [c]
 SET [c].[ContactName] = N'Updated'
 FROM [Customers] AS [c]
-WHERE [c].[CustomerID] LIKE N'F%'
+INNER JOIN (
```
Comment (roji) [inline test/EFCore.SqlServer.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesSqlServerTest.cs:972]: This SQL becomes seemingly more complex. However, the simpler SQL was only the result of logic for ignoring Distinct in cases where it's totally useless (i.e. when placed direct on entity types which have unique PKs, so necessarily a no-op). which doesn't fit so well with the new code. The user can simply remove their useless Distinct() if they want.

File: src/EFCore.Relational/Query/Internal/TpcTablesExpression.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+namespace Microsoft.EntityFrameworkCore.Query.Internal;
+
+/// <summary>
+///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+///     any release. You should only use it directly in your code with extreme caution and knowing that
+///     doing so can result in application failures when updating to a new Entity Framework Core release.
+/// </summary>
+public sealed class TpcTablesExpression : TableExpressionBase
```
Comment (roji) [inline src/EFCore.Relational/Query/Internal/TpcTablesExpression.cs:15]: Need to expose it in order to recognize it in ExecuteUpdate translation, to throw the appropriate error message; it's still internal (another part of opening things up a bit).

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -1662,70 +1522,195 @@ static bool AreOtherNonOwnedEntityTypesInTheTable(IEntityType rootType, ITableBa
...
+            {
+                AddTranslationErrorDetails(
+                    RelationalStrings.ExecuteOperationOnKeylessEntityTypeWithUnsupportedOperator(
+                        nameof(RelationalQueryableExtensions.ExecuteUpdate),
+                        entityType.DisplayName()));
+                return null;
+            }
+
+            // Generate the INNER JOIN around the original query, on the PK properties.
+            var outer = (ShapedQueryExpression)Visit(new EntityQueryRootExpression(entityType));
```
Comment (roji) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1602]: Most of these changes are simply refactoring/moving around the code.

File: test/EFCore.SqlServer.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesSqlServerTest.cs (hunk 1 above)
Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/BulkUpdates/NorthwindBulkUpdatesSqlServerTest.cs:972]: BTW thinking about this, if we really want to remove useless Distincts, that should probably be done more generally in the query pipeline (though again, it's the kind of optimization which only speeds up badly-written LINQ queries).

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -1440,194 +1437,57 @@ static bool AreOtherNonOwnedEntityTypesInTheTable(IEntityType rootType, ITableBa
...
-                AddTranslationErrorDetails(RelationalStrings.InvalidPropertyInSetProperty(propertyExpression.Print()));
+                AddTranslationErrorDetails(RelationalStrings.InvalidPropertyInSetProperty(propertySelector.Print()));
                 return null;
             }
 
-            if (entityShaperExpression is null)
+            if (targetTable is null)
             {
-                entityShaperExpression = ese;
+                (targetTable, targetTablePropertySelector) = (column.Table, propertySelector);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1462]: Use separate statements here.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -1662,70 +1522,195 @@ static bool AreOtherNonOwnedEntityTypesInTheTable(IEntityType rootType, ITableBa
...
-                if (mce.TryGetIndexerArguments(model, out var source2, out _))
+                columns[i] = column;
+            }
+
+            return TranslateValueExpressions(this, outer, outerSelectExpression, tableExpression, propertyValueLambdaExpressions, columns);
+
+            // For property setter selectors in ExecuteUpdate, we support only simple member access, EF.Function, etc.
+            // We also unwrap casts to interface/base class (#29618). Note that owned IncludeExpressions have already been pruned from the
+            // source before remapping the lambda (#28727).
+            bool TryExtractEntityType(Expression expression, IModel model, [NotNullWhen(true)] out IEntityType? entityType)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1667]: Avoid nested local functions when possible.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 6]
```diff
@@ -1787,43 +1772,33 @@ static Expression GetEntitySource(IModel model, Expression propertyAccessExpress
     ///     </para>
     /// </remarks>
     /// <param name="selectExpression">The select expression to validate.</param>
-    /// <param name="entityShaperExpression">The entity shaper expression on which the update operation is being applied.</param>
+    /// <param name="table">TODO</param>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1775]: 👀

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 7]
```diff
@@ -1787,43 +1772,33 @@ static Expression GetEntitySource(IModel model, Expression propertyAccessExpress
...
+            })
+        {
+            if (selectExpression.Tables.Count > 1)
             {
-                var projectionBindingExpression = (ProjectionBindingExpression)entityShaperExpression.ValueBufferExpression;
-                var entityProjectionExpression = (EntityProjectionExpression)selectExpression.GetProjection(projectionBindingExpression);
-                var column = entityProjectionExpression.BindProperty(entityShaperExpression.EntityType.GetProperties().First());
-                table = column.Table;
-                if (ReferenceEquals(selectExpression.Tables[0], table))
+                // If the table we are looking for it first table, then we need to verify if we can lift the next table in FROM clause
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1797]:
```suggestion
                // If the table we are looking for is the first table, then we need to verify whether we can lift the next table in FROM clause
```

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**